### PR TITLE
New resources: aws_dx_gateway, aws_dx_gateway_assocation

### LINKED ISSUE #2140: Add support for Direct Connect Gateway
AWS just released Direct Connect Gateway which enable cross-region VPC communication. Filing this issue to track the feature in the terraform AWS provider.

https://aws.amazon.com/blogs/aws/new-aws-direct-connect-gateway-inter-region-vpc-access/

### PR DESCRIPTION
**Pull request by aeriff**
```
TF_ACC=1 go test ./aws/ -v -run=TestAccAwsDxGateway -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_basic
--- PASS: TestAccAwsDxGatewayAssociation_basic (65.53s)
=== RUN   TestAccAwsDxGatewayAssociation_multiVgws
--- PASS: TestAccAwsDxGatewayAssociation_multiVgws (79.58s)
=== RUN   TestAccAwsDxGateway_basic
--- PASS: TestAccAwsDxGateway_basic (31.14s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	176.300s
```

I have used the same ASN validation code from #1888 to prove this works, but I suppose this blocks merging this PR until that is solved with the proposed addition of `TypeInt64` to Terraform.

Resolves #2140.

### CODE REVIEW

**Comment by Ninir** [comment]
Hi @aeriff 

Thanks for the work here, looks promising! Would you mind splitting this PR in two (if possible) with 1 resource per PR?

Would be easier for us to review & merge :)

Thanks!

**Comment by aeriff** [comment]
@Ninir thanks for the feedback. I have split them into separate PRs as requested.

File: aws/resource_aws_dx_gateway.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsDxGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsDxGatewayCreate,
+		Read:   resourceAwsDxGatewayRead,
+		Delete: resourceAwsDxGatewayDelete,
+		Importer: &schema.ResourceImporter{
```
Comment (ewbankkit) [inline aws/resource_aws_dx_gateway.go]: If `terraform import` is supported please note in the documentation and add an import acceptance test.
I did that for the existing Direct Connect resources in https://github.com/terraform-providers/terraform-provider-aws/pull/2992.

**Comment by aeriff** [comment]
@ewbankkit thanks for the reminder! Adding import support was an afterthought, so I missed writing the tests.

```
TF_ACC=1 go test ./aws -v -run=TestAccAwsDxGateway_import -timeout 120m
=== RUN   TestAccAwsDxGateway_importBasic
--- PASS: TestAccAwsDxGateway_importBasic (36.99s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	37.029s
```~

**Comment by aeriff** [comment]
I decided to merge the 2 PRs back into one because both resources now have dependencies on each other. The resources are in 2 separate commits, so review shouldn't be any more difficult. All the discussion has happened so far in this PR, so I will close #2862 in favour of this. 

Acceptance tests:
```
TF_ACC=1 go test ./aws -v -run=TestAccAwsDxGateway -timeout 120m
=== RUN   TestAccAwsDxGateway_importBasic
--- PASS: TestAccAwsDxGateway_importBasic (36.29s)
=== RUN   TestAccAwsDxGateway_importComplex
--- PASS: TestAccAwsDxGateway_importComplex (116.35s)
=== RUN   TestAccAwsDxGatewayAssociation_basic
--- PASS: TestAccAwsDxGatewayAssociation_basic (98.23s)
=== RUN   TestAccAwsDxGatewayAssociation_multiVgws
--- PASS: TestAccAwsDxGatewayAssociation_multiVgws (124.51s)
=== RUN   TestAccAwsDxGateway_basic
--- PASS: TestAccAwsDxGateway_basic (31.80s)
PASS
ok	github.com/terraform-providers/terraform-provider-aws/aws	407.018s
```

**Comment by aeriff** [comment]
@radeksimko thanks for the review! All comments should all be addressed now.

```
=== RUN   TestAccAwsDxGateway_importBasic
--- PASS: TestAccAwsDxGateway_importBasic (48.41s)
=== RUN   TestAccAwsDxGateway_importComplex
--- PASS: TestAccAwsDxGateway_importComplex (77.33s)
=== RUN   TestAccAwsDxGatewayAssociation_basic
--- PASS: TestAccAwsDxGatewayAssociation_basic (78.88s)
=== RUN   TestAccAwsDxGatewayAssociation_multiVgws
--- PASS: TestAccAwsDxGatewayAssociation_multiVgws (123.67s)
=== RUN   TestAccAwsDxGateway_basic
--- PASS: TestAccAwsDxGateway_basic (43.70s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	372.046s
```

The tests pass for me (almost) consistently. I also saw one `IncorrectState` error during one run when trying to delete a test VGW, but I would argue that's out of scope of these changes and more likely a bug in the retry logic in the `aws_vpn_gateway` resource.

I guess merging is blocked by hashicorp/terraform#17438, however. I tested with a custom random range function for now.

**Comment by kmcquade** [comment]
@aeriff any idea when this will be released? Just checking. Thanks for your work!

**Comment by MrGossett** [comment]
It would be great to see this merged.

**Comment by DevilWAH** [comment]
Has this been merged yet? being able to associate a direct connect gateway to VPC gateway would be really nice :)

**Comment by violetmythmaker** [comment]
Definitely looking forward to seeing this merged. Will we be able to pass the ID of the created gateway to a route's `gateway_id` argument as well, like with an `aws_internet_gateway` or `aws_vpn_gateway` ?

**Comment by sleterrier** [comment]
While seeing this merged would be great, I wonder how viable it would be in a complex environment with multiple regions and accounts. Is there any plan to add a aws_dx_gateway data source at some point?

**Comment by ewbankkit** [comment]
@sleterrier Please open an issue(s) for any data source(s) you'd like added.

**Comment by ewbankkit** [comment]
@violetmythmaker The association between a Direct Connect Gateway and a Virtual Private Gateway is managed via the `aws_dx_gateway_association` resource added in this PR.
I'm not aware of the ability to associate a Direct Connect Gateway with an Internet Gateway.

**Comment by heycasey** [comment]
Also looking forward to seeing this merged. Are there any additional steps needed before the merge

@ewbankkit, you are correct that a DX gateway cannot be associated with an Internet Gateway, only VGWs.

**Comment by ewbankkit** [comment]
@aeriff If you added your custom `RandIntRange` function then acceptance tests will pass and this should aid in getting the PR merged.
Once https://github.com/hashicorp/terraform/pull/17438 is merged we can change back.

Also, for https://github.com/terraform-providers/terraform-provider-aws/pull/3253 and https://github.com/terraform-providers/terraform-provider-aws/pull/3255 I use the name `vpn_gateway_id` for the attribute similar to the one you have named `virtual_gateway_id` in the new `aws_dx_gateway_association` resource. The attributes should have the same name.
I have no problem changing the name in my 2 PRs.
I named the attribute `vpn_gateway_id` as that is what the attribute is [named](https://www.terraform.io/docs/providers/aws/r/vpn_gateway_attachment.html#vpn_gateway_id) for its association with a VPC in the `aws_vpn_gateway_attachment` resource.

**Comment by ewbankkit** [comment]
@aeriff I took the liberty of making the suggested changes in https://github.com/terraform-providers/terraform-provider-aws/pull/4896 based off this PR. The acceptance tests are now passing.

**Comment by aeriff** [comment]
@ewbankkit thanks! Closing this in favour of your PR.

**Pull request closed without merging**